File: src/main/java/com/timepath/steam/io/storage/VPK.java
package com.timepath.steam.io.storage;

import com.timepath.DataUtils;
import com.timepath.StringUtils;
import com.timepath.io.ByteBufferInputStream;
import com.timepath.io.struct.StructField;
import com.timepath.steam.io.storage.Files.FileHandler;
import com.timepath.steam.io.util.ExtendedVFile;
import com.timepath.vfs.SimpleVFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.FileFilter;
import java.io.IOException;
import java.io.InputStream;
import java.lang.ref.Reference;
import java.lang.ref.SoftReference;
import java.nio.ByteBuffer;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.zip.CRC32;
import java.util.zip.Checksum;

/**
 * Loads _dir.vpk files
 *
 * @author TimePath
 * @see <a>https://developer.valvesoftware.com/wiki/VPK_File_Format</a>
 */
public class VPK extends ExtendedVFile {

    private static final int HEADER = 0x55AA1234;
    private static final Logger LOG = Logger.getLogger(VPK.class.getName());
    /**
     * Previously loaded VPKs stored as references.
     */
    private static final Map<File, Reference<VPK>> REFERENCE_MAP = new HashMap<>(0);
    private final ByteBuffer globaldata;
    @NotNull
    private final ByteBuffer[] mappings;
    @NotNull
    private final File[] store;
    private String name;

    static {
        Files.registerHandler(new FileHandler() {
            @Nullable
            @Override
            public Collection<? extends SimpleVFile> handle(@NotNull final File file) throws IOException {
                if (!file.getName().endsWith("_dir.vpk")) return null;
                return VPK.loadArchive(file).list();
            }
        });
    }

    private VPK(@NotNull final File file) throws IOException {
        LOG.log(Level.INFO, "Loading {0}", file);
        // Map extra archives
        name = file.getName();
        name = name.substring(0, name.length() - 4); // Strip '.vkp'
        if (name.endsWith("_dir")) {
            name = name.substring(0, name.length() - 4); // Strip '_dir'
        }
        File[] parts = file.getParentFile().listFiles(new FileFilter() {
            @Override
            public boolean accept(@NotNull File pathname) {
                return !pathname.equals(file) && pathname.getName().startsWith(name) &&
                        (pathname.getName().length() == (name.length() + 8));
            }
        });
        store = new File[parts.length];
        mappings = new ByteBuffer[store.length];
        for (@NotNull File part : parts) {
            @NotNull String[] split = part.getName().split("_");
            int idx = Integer.parseInt(split[split.length - 1].replaceAll(".vpk", ""));
            store[idx] = part;
        }
        ByteBuffer buffer = DataUtils.mapFile(file);
        int signature = buffer.getInt();
        if (signature != HEADER) {
            throw new IOException("Not a VPK file");
        }
        int ver = buffer.getInt();
        int treeLength = buffer.getInt(); // Unsigned length of directory slice
        int dataLength = 0;
        int v2 = 0;
        int v3 = 0; // 48 in most
        int v4 = 0;
        if (ver >= 2) {
            dataLength = buffer.getInt();
            v2 = buffer.getInt();
            v3 = buffer.getInt();
            v4 = buffer.getInt();
        }
        ByteBuffer directoryInfo = DataUtils.getSlice(buffer, treeLength);
        globaldata = DataUtils.getSlice(buffer, dataLength);
        buffer.get(new byte[v2]); // Directory
        buffer.get(new byte[v3]); // Single + Directory
        buffer.get(new byte[v4]); // Directory
        @NotNull Object[][] debug = {
                {"dataLength = ", dataLength},
                {"v2 = ", v2},
                {"v3 = ", v3},
                {"v4 = ", v4},
                {"Underflow = ", buffer.remaining()},
        };
        LOG.info(StringUtils.fromDoubleArray(debug, "Debug:"));
        parseTree(directoryInfo);
    }

    @Nullable
    public static VPK loadArchive(@NotNull File file) {
        Reference<VPK> ref = REFERENCE_MAP.get(file);
        @Nullable VPK cached = (ref != null) ? ref.get() : null;
        if (cached != null) {
            LOG.log(Level.INFO, "Loaded {0} from cache", file);
            return cached;
        }
        try {
            @NotNull VPK v = new VPK(file);
            REFERENCE_MAP.put(file, new SoftReference<>(v));
            return v;
        } catch (IOException ex) {
            LOG.log(Level.SEVERE, null, ex);
            return null;
        }
    }

    private void parseTree(@NotNull ByteBuffer buffer) {
        // Extensions
        for (String ext; !(ext = DataUtils.readZeroString(buffer)).isEmpty(); ) {
            ext = " ".equals(ext) ? "" : '.' + ext;
            // Paths
            for (String dir; !(dir = DataUtils.readZeroString(buffer)).isEmpty(); ) {
                @Nullable SimpleVFile p = nodeForPath(dir);
                // File names
                for (String basename; !(basename = DataUtils.readZeroString(buffer)).isEmpty(); ) {
                    @NotNull VPKDirectoryEntry e = readFileInfo(buffer, basename + ext);
                    p.add(e);
                }
            }
        }
    }

    @Nullable
    private SimpleVFile nodeForPath(@NotNull String path) {
        @Nullable SimpleVFile node = getRoot();
        if (" ".equals(path)) {
            return node;
        }
        @NotNull String[] components = path.replace('\\', '/').split("/");
        for (String dir : components) {
            @Nullable SimpleVFile match = null;
            for (@NotNull SimpleVFile e : node.list()) {
                if (e.isDirectory() && e.getName().equalsIgnoreCase(dir)) {
                    match = e;
                    break;
                }
            }
            if (match == null) {
                @NotNull VPKDirectoryEntry dirEntry = new VPKDirectoryEntry(dir);
                dirEntry.isDirectory = true;
                match = dirEntry;
                node.add(match);
            }
            node = match;
        }
        return node;
    }

    @NotNull
    private VPKDirectoryEntry readFileInfo(@NotNull ByteBuffer buffer, String name) {
        @NotNull VPKDirectoryEntry e = new VPKDirectoryEntry(name);
        e.crc = buffer.getInt();
        e.preloadBytes = buffer.getShort();
        e.archiveIndex = buffer.getShort();
        e.entryOffset = buffer.getInt();
        e.entryLength = buffer.getInt();
        buffer.position(buffer.position() + e.preloadBytes); // TODO: load preload bytes
        int term = buffer.getShort() & 0xFFFF;
        assert term == 0xFFFF : "VPK directory reading failed";
        return e;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public boolean isDirectory() {
        return true;
    }

    @Nullable
    private ByteBuffer getData(int i) {
        try {
            if (mappings[i] == null) {
                mappings[i] = DataUtils.mapFile(store[i]);
            }
            return mappings[i];
        } catch (IOException ex) {
            LOG.log(Level.SEVERE, null, ex);
        }
        return null;
    }

    @Nullable
    @Override
    public Object getAttributes() {
        return null;
    }

    @NotNull
    @Override
    public ExtendedVFile getRoot() {
        return this;
    }

    /**
     * VPK entries are complete by definition
     *
     * @return true
     */
    @Override
    public boolean isComplete() {
        return true;
    }

    @Nullable
    @Override
    public InputStream openStream() {
        return null;
    }

    /**
     * If a file contains preload data, the preload data immediately follows the
     * above structure. The entire size of a file is PreloadBytes + EntryLength.
     */
    private class VPKDirectoryEntry extends ExtendedVFile {

        /**
         * A 32bit CRC of the file's data.
         */
        @StructField(index = 0)
        int crc;
        @StructField(index = 2)
        short archiveIndex;
        @StructField(index = 4)
        int entryLength;
        @StructField(index = 3)
        int entryOffset;
        boolean isDirectory;
        String name;
        @StructField(index = 1)
        short preloadBytes;
        @Nullable
        private Reference<ByteBuffer> localdata;

        VPKDirectoryEntry(String name) {
            this.name = name;
        }

        @Override
        public long calculateChecksum() {
            if (localData() == null) {
                return 0;
            }
            @NotNull Checksum checksum = new CRC32();
            localData().position(0);
            @NotNull byte[] buf = new byte[4096];
            while (localData().hasRemaining()) {
                int bsize = Math.min(buf.length, localData().remaining());
                localData().get(buf, 0, bsize);
                checksum.update(buf, 0, bsize);
            }
            return checksum.getValue();
        }

        @Nullable
        @Override
        public Object getAttributes() {
            return null;
        }

        @Override
        public long getChecksum() {
            return crc;
        }

        @Override
        public String getName() {
            return name;
        }

        @NotNull
        @Override
        public ExtendedVFile getRoot() {
            return VPK.this;
        }

        @Nullable
        public ByteBuffer getSource() {
            if (archiveIndex == 0x7FFF) { // This archive
                return globaldata;
            }
            return getData(archiveIndex);
        }

        @Override
        public boolean isComplete() {
            long theoretical = crc;
            long real = calculateChecksum();
            return theoretical == real;
        }

        @Override
        public boolean isDirectory() {
            return isDirectory;
        }

        @Override
        public long length() {
            return entryLength;
        }

        @Nullable
        public ByteBuffer localData() {
            @Nullable ByteBuffer buf = (localdata != null) ? localdata.get() : null;
            if (buf == null) {
                @Nullable ByteBuffer src = getSource();
                src.position(entryOffset);
                buf = DataUtils.getSlice(src, entryLength);
                localdata = new SoftReference<>(buf);
            }
            return buf;
        }

        @Nullable
        @Override
        public InputStream openStream() {
            return new ByteBufferInputStream(localData());
        }
    }
}

File: src/main/kotlin/com/timepath/steam/io/storage/VPK.kt
package com.timepath.steam.io.storage

import com.timepath.DataUtils
import com.timepath.StringUtils
import com.timepath.io.ByteBufferInputStream
import com.timepath.io.struct.StructField
import com.timepath.toUnsigned
import com.timepath.vfs.SimpleVFile
import com.timepath.vfs.provider.ExtendedVFile
import com.timepath.with
import java.io.File
import java.io.FileFilter
import java.io.IOException
import java.lang.ref.Reference
import java.lang.ref.SoftReference
import java.nio.ByteBuffer
import java.util.HashMap
import java.util.logging.Level
import java.util.logging.Logger
import java.util.zip.CRC32

/**
 * Loads _dir.vpk files
 * @see [https://developer.valvesoftware.com/wiki/VPK_File_Format](null)
 */
public class VPK @throws(IOException::class)
private constructor(file: File) : ExtendedVFile() {
    override val root = this

    private val globaldata: ByteBuffer
    private val mappings: Array<ByteBuffer?>
    private val store: Array<File?>
    override val name: String

    init {
        LOG.log(Level.INFO, "Loading {0}", file)
        name = file.getName()
                .let { it.substringBeforeLast(".vpk", it) }
                .let { it.substringBeforeLast("_dir", it) }
        val parts = file.getParentFile().listFiles(FileFilter {
            it != file
                    && it.getName().startsWith(name)
                    && it.getName().length() == name.length() + "_000.vpk".length()
        })
        store = arrayOfNulls<File>(parts.size())
        mappings = arrayOfNulls<ByteBuffer>(store.size())
        for (part in parts) {
            val split = part.getName().splitBy("_")
            val idx = Integer.parseInt(split[split.size() - 1].replace(".vpk", ""))
            store[idx] = part
        }
        val buffer = DataUtils.mapFile(file)
        val signature = buffer.getInt()
        if (signature != HEADER) {
            throw IOException("Not a VPK file")
        }
        val ver = buffer.getInt()
        val treeLength = buffer.getInt() // Unsigned length of directory slice
        val dataLength: Int
        val v2: Int
        val v3: Int // 48 in most
        val v4: Int
        if (ver >= 2) {
            dataLength = buffer.getInt()
            v2 = buffer.getInt()
            v3 = buffer.getInt()
            v4 = buffer.getInt()
        } else {
            dataLength = 0
            v2 = 0
            v3 = 0
            v4 = 0
        }
        val directoryInfo = DataUtils.getSlice(buffer, treeLength)
        globaldata = DataUtils.getSlice(buffer, dataLength)
        buffer.get(ByteArray(v2)) // Directory
        buffer.get(ByteArray(v3)) // Single + Directory
        buffer.get(ByteArray(v4)) // Directory
        val debug = arrayOf(
                arrayOf<Any>("dataLength = ", dataLength),
                arrayOf<Any>("v2 = ", v2),
                arrayOf<Any>("v3 = ", v3),
                arrayOf<Any>("v4 = ", v4),
                arrayOf<Any>("Underflow = ", buffer.remaining())
        )
        LOG.info(StringUtils.fromDoubleArray(debug, "Debug:"))
        parseTree(directoryInfo)
    }

    private inline fun parse(buffer: ByteBuffer, body: (String) -> Unit) {
        while (true) {
            val it = DataUtils.readZeroString(buffer)
            if (it.isEmpty()) break
            body(it)
        }
    }

    private fun parseTree(buffer: ByteBuffer) {
        parse(buffer) {
            val ext = when (it) {
                " " -> ""
                else -> '.' + it
            }
            parse(buffer) { dir ->
                val p = nodeForPath(dir)
                parse(buffer) { basename ->
                    val e = readFileInfo(buffer, basename + ext)
                    p.add(e)
                }
            }
        }
    }

    private fun nodeForPath(path: String): SimpleVFile {
        if (path == " ") return root
        return path.replace('\\', '/').splitBy("/").fold(root as SimpleVFile) { parent, child ->
            parent.list().firstOrNull {
                it.isDirectory && it.name.equals(child, ignoreCase = true)
            } ?: run {
                VPKDirectoryEntry(child) with {
                    isDirectory = true
                    parent.add(this)
                }
            }
        }
    }

    private fun readFileInfo(buffer: ByteBuffer, name: String): VPKDirectoryEntry {
        val e = VPKDirectoryEntry(name) with {
            crc = buffer.getInt()
            preloadBytes = buffer.getShort()
            archiveIndex = buffer.getShort()
            entryOffset = buffer.getInt()
            entryLength = buffer.getInt()
        }
        buffer.position(buffer.position() + e.preloadBytes) // TODO: load preload bytes
        val term = buffer.getShort().toUnsigned()
        assert(term == 0xFFFF, "VPK directory reading failed")
        return e
    }

    override val isDirectory = true

    private fun getData(i: Int): ByteBuffer? {
        try {
            return mappings[i] ?: DataUtils.mapFile(store[i]!!) with { mappings[i] = this }
        } catch (ex: IOException) {
            LOG.log(Level.SEVERE, null, ex)
        }

        return null
    }

    override val attributes = null

    override val isComplete = true

    override fun openStream() = null

    /**
     * If a file contains preload data, the preload data immediately follows the
     * above structure. The entire size of a file is PreloadBytes + EntryLength.
     */
    private inner class VPKDirectoryEntry(override var name: String) : ExtendedVFile() {

        /**
         * A 32bit CRC of the file's data.
         */
        StructField(0) var crc: Int = 0
        StructField(1) var preloadBytes: Short = 0
        StructField(2) var archiveIndex: Short = 0
        StructField(3) var entryOffset: Int = 0
        StructField(4) var entryLength: Int = 0

        override var isDirectory: Boolean = false
        private var localdata: Reference<ByteBuffer>? = null

        override fun calculateChecksum(): Long {
            if (localData() == null) {
                return 0
            }
            val checksum = CRC32()
            localData()!!.position(0)
            val buf = ByteArray(4096)
            while (localData()!!.hasRemaining()) {
                val bsize = Math.min(buf.size(), localData()!!.remaining())
                localData()!!.get(buf, 0, bsize)
                checksum.update(buf, 0, bsize)
            }
            return checksum.getValue()
        }

        override val attributes = null

        override val checksum: Long get() = crc.toLong()

        override val root = this@VPK

        public fun getSource(): ByteBuffer? {
            if (archiveIndex.toInt() == 0x7FFF) {
                // This archive
                return globaldata
            }
            return getData(archiveIndex.toInt())
        }

        override val isComplete: Boolean get() {
            val theoretical = crc.toLong()
            val real = calculateChecksum()
            return theoretical == real
        }

        override val length: Long get() = entryLength.toLong()

        public fun localData(): ByteBuffer? {
            localdata?.get()?.let { return it }
            return DataUtils.getSlice(getSource()!!.with { position(entryOffset) }, entryLength) with {
                localdata = SoftReference(this)
            }
        }

        override fun openStream() = ByteBufferInputStream(localData()!!)
    }

    companion object {

        private val HEADER = 0x55AA1234
        private val LOG = Logger.getLogger(javaClass<VPK>().getName())
        /**
         * Previously loaded VPKs stored as references.
         */
        private val REFERENCE_MAP = HashMap<File, Reference<VPK>>(0)

        public fun loadArchive(file: File): VPK? {
            REFERENCE_MAP[file]?.get()?.let {
                LOG.log(Level.INFO, "Loaded {0} from cache", file)
                return it
            }
            try {
                return VPK(file) with { REFERENCE_MAP[file] = SoftReference(this) }
            } catch (ex: IOException) {
                LOG.log(Level.SEVERE, null, ex)
                return null
            }
        }
    }
}
